### PR TITLE
Space - Olivia Mulholland Salazar

### PR DESCRIPTION
**Pull request by mulhoo**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | The hardest part was getting everything to connect properly, as well as figuring out how to navigate the various layers. I had a hard time grasping classes and building them on my own, so this project was hard for me with needing to completely build my own and not having pre-written tests there to guide me if I got it or not. Getting my room assignment code to all connect properly is something I am still working on. The individual pieces work how they should, but the unit doesn't quite work yet, so expect updates!!
What was a design decision you made that changed over time over the project? | The biggest addition was adding in the calendar_date class to manage each day's reservations. Originally I was going to have calendar do everything, but that ended up being very complicated so I created another class to manage the individual dates.
What was a concept you gained clarity on, or a learning that you'd like to share? | I feel like my understand of classes and their relationship to one another as a whole improved a lot, but I still need to work on doing my tests first rather than my code. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | Ensuring that the calculated cost of the reservation is correct. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | Haven't explicitly written the test yet since I've been so focused on getting the room assignment to work, but throwing up an error when the date isn't inputted properly, or is completely invalid. I also plan to write one for if the room number is negative for whatever reason as well as a few others in other places. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | The pseudocode was really helpful, as I had a very clear picture and plan for what I wanted to do and how I wanted everything to work. It definitely helped a lot to keep me organized, especially now that I am struggling to find where my code is not working 100% properly when it comes to assigning rooms to reservations.

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -3,7 +3,7 @@ require 'rake/testtask'
 Rake::TestTask.new do |t|
   t.libs = ["lib"]
   t.warning = true
-  t.test_files = FileList['test/*_test.rb']
+  t.test_files = FileList['test/*_spec.rb']
```
Comment (dHelmgren) [inline Rakefile:6]: Rails will soon start enforcing `_test` as the file ending, so it's better to start conforming your file names to what's in the Rakefile.

File: lib/calendar.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    return reservation
+  end
+
+
+  def list_available_rooms_entire_stay(reservation)
+    date_range = []
+    available_rooms_all_dates = []
+
+
+    if (reservation.total_nights == 1) && (@date_store[(reservation.start_date).iso8601].all_available_rooms.length == 20)
```
Comment (dHelmgren) [inline lib/calendar.rb:27]: what is `iso8601`? Is this necessary?

File: lib/calendar.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      available_rooms = [1]
+      return available_rooms
+    end
+
+    (reservation.total_nights).times do |i|
+      if @date_store[(reservation.start_date + i).iso8601].nil?
+        date = CalendarDate.new
```
Comment (dHelmgren) [inline lib/calendar.rb:34]: You make a new `CalendarDate` but your `date` variable doesn't leave the `if` block. What was your goal here?

File: lib/calendar.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      else
+        raise NoRoomsError.new "There are no rooms available for the duration of this stay." if @date_store[(reservation.start_date + i).iso8601].unavailable?
+        date_range << (@date_store[(reservation.start_date + i).iso8601]).all_available_rooms
+      end
+
+    end
+
+
+    date_range.each do |i|
```
Comment (dHelmgren) [inline lib/calendar.rb:43]: What you want here is `each_index`. By default `.each` looks at the elements themselves.

File: lib/calendar.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    raise NoRoomsError.new "There are no rooms available for the duration of this stay." if available_rooms_all_dates.empty?   
+
+    available_rooms = available_rooms_all_dates.flatten.uniq!
+
+    return available_rooms
+  end
+
+
+  def is_available?(room_number, date)
+    date = Date.parse(date).iso8601
```
Comment (dHelmgren) [inline lib/calendar.rb:58]: Not sure why you're doing `parse` here. You can build the software to enforce that idea that the date is already parsed when it was given to you. 

When something feels like more work than it ought to be, that's a good time to ask questions! Asking lots of questions about the scope of a project will serve you well at Ada and beyond!

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      true
+    rescue ArgumentError
+      false
+    end
+
+    valid_date?(start_date)
+    valid_date?(end_date)
+
+    raise InvalidDateRangeError.new ("The given date range is invalid.") if (@end_date - @start_date) <= 0
+  end
```
Comment (dHelmgren) [inline lib/reservation.rb:30]: Looks like you defined a method inside a method, which Ruby REALLY doesn't like.

```
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: method redefined; discarding old valid_date?
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: previous definition of valid_date? was here
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: method redefined; discarding old valid_date?
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: previous definition of valid_date? was here
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: method redefined; discarding old valid_date?
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: previous definition of valid_date? was here
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: method redefined; discarding old valid_date?
/Users/devin/Documents/Ada/c13/hotel/lib/reservation.rb:18: warning: previous definition of valid_date? was here
```

File: test/calendar_date_spec.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    @calendar = Calendar.new
+    
+    @calendar.add_to_calendar(@reservation1)
+    @calendar.add_to_calendar(@reservation2)
+    @calendar.add_to_calendar(@reservation3)
+  end
+
+  it "has an array that keeps track of the available rooms" do
+    
+    hopefully_an_array = @date_store[Date.parse("2020/05/04").iso8601].all_available_rooms
```
Comment (dHelmgren) [inline test/calendar_date_spec.rb:18]: I think you meant to write this:
```suggestion
    hopefully_an_array = @calendar.date_store[Date.parse("2020/05/04").iso8601].all_available_rooms
```

File: test/calendar_date_spec.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    expect hopefully_an_array.must_be_instance_of Array
```
Comment (dHelmgren) [inline test/calendar_date_spec.rb:20]: Be sure to use parens with `expect`, in situations like this it might not do what you think it's doing.
```suggestion
    expect(hopefully_an_array).must_be_instance_of Array
```

File: test/calendar_spec.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    # puts "\n"
+    
+  end
+
+  it "gives correct cost of requested reservation" do
+    reservation1 = @reservation1.details
+    id = reservation1[:id]
+    cost = @calendar.total_cost_reservation(id)
+
+    expect (reservation1[:cost]).must_equal cost
```
Comment (dHelmgren) [inline test/calendar_spec.rb:27]: Putting a space here means `expect` is called without arguments, which will cause your tests to misbehave.
```suggestion
    expect(reservation1[:cost]).must_equal cost
```

File: test/calendar_spec.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+  it 'creates a new reservation' do
+    expected_outcome = Reservation.new("2020/05/04", "2020/05/07")
+    actual_outcome = expected_outcome
+
+    assert_equal(actual_outcome, expected_outcome)
+  end
+
+  it "returns true/false if given room is available on a given date" do
+    yesorno = @calendar.is_available?(5, "2020/05/07")
```
Comment (dHelmgren) [inline test/calendar_spec.rb:76]: Snake case!
```suggestion
    yes_or_no = @calendar.is_available?(5, "2020/05/07")
```

File: test/calendar_spec.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    cost = @calendar.total_cost_reservation(id)
+
+    expect (reservation1[:cost]).must_equal cost
+  end
+
+  it "lists all reservations for specific date" do
+    reservation1 = @reservation1.details
+    expected_outcome = [reservation1]
+    actual_outcome = @calendar.reservations_on_date("2020/05/04")
+    assert_equal(actual_outcome, expected_outcome)
```
Comment (dHelmgren) [inline test/calendar_spec.rb:34]: With a test case of only one, this test is a little inconclusive.

**Comment by dHelmgren** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert |
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | (lots of code w/o tests)
Practices git with at least 15 small commits and meaningful commit messages |

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | (Looks like this is being handled entirely by the calendar right now)
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests |
All of the given tests run and pass |
A test coverage tool is installed and used, and shows 95% test coverage |

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range |
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date |
calculates the total price for a reservation |
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range |
creates a block of rooms |
reserves a room from a block |

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections |
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | ✔️

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you approached the challenge of design with a unique plan. It's challenging to build a design from the ground up, especially when there are such complicated needs embedded in the problem.

I do see some room for improvement around getting guidance earlier and improving you comfort with writing tests. Your ruby code is generally well organized into methods and very readable, but it's clear to me that at some point the project got a little bit away from you. In situations like that (they happen in industry too), the best thing to do is sit down with someone who is more experienced and say as much. Whomever they are can help you prune problem code and get you back on the rails. Second, I see a couple of syntactic slips here and there inside and outside your tests that are causing problems. By focusing on doing Red Green Refactor, you can get a lot clearer picture of how and when things went wrong before problems really start to pile up. An important part of this is making sure your tests fail before you expect them to pass.

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
